### PR TITLE
[docs] start a new FSDP notes doc

### PR DESCRIPTION
**Pull request by stas00**
As discussed on [slack](https://pytorch.slack.com/archives/C3PDTEV8E/p1703699711772289) adding Andrew Gu's advanced FSDP design notes with a few additions from myself based on our discussion.

I hope I did the RST right, I haven't done RST in a while.

- The first section is Andrew's words verbatim + formatting
- The second section is Andrew's words verbatim + formatting + a few of my additions that were confirmed by Andrew, and which hopefully should help understand the process better.

tagging @albanD as requested.

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @fegin @XilunWu @wanchaol @fduwjj @wz337 @tianyu-l @wconstab @yf225

### CODE REVIEW

File: docs/source/notes/fsdp.rst
Comment (awgu) [inline docs/source/notes/fsdp.rst:74]: This statement that `forward_prefetch=True` needs 2x buffer size is correct. However, `forward_prefetch=False` _also_ needs 2x buffer size.

The reason is that in the flat-parameter FSDP design, we _do not_ copy-out of the all-gather buffer. The parameters used for compute are directly viewed into the all-gather buffer (in fact, the main benefit of the "flat parameter" is exactly this reason). In that case, while 'layer 1 all-gather' is overlapping with 'layer 0 forward compute', the 'layer 0 forward compute' is using the parameters viewed into the 'layer 0 all-gather' buffer.

A natural question then is, when would you want `forward_prefetch=False`? For static-graph models (like most LLMs), I do not think that there is a major technical reason. It is more that, practically, we added this option quickly for some CPU-bound internal models and have not tested every code path with it in unit testing, so we are less confident in it. For example, we found a bug just recently (https://github.com/pytorch/pytorch/pull/116792). `forward_prefetching=False` can be slightly easier to reason about since we do not have to check the recorded forward order as a possible 'failure mode'; a module's all-gather can always be found under its own `record_function` label in its profiler trace.

File: docs/source/notes/fsdp.rst
Comment (awgu) [inline docs/source/notes/fsdp.rst:58]: nit (typo):
```suggestion
additional communication since the weights are prefetched anyway during ``backward``.
```

File: docs/source/notes/fsdp.rst
Comment (awgu) [inline docs/source/notes/fsdp.rst:53]: nit (suggestion): can change "weights" -> "parameters" following PyTorch convention

**Review by awgu — COMMENTED** [review]
The docs render looks good to me (so you did the `.rst` properly 😄 ).

Thanks so much for taking the time to add this! I left one main comment and a few nits.

**Comment by awgu** [comment]
Could you also run `lintrunner -a` to lint when you get the chance? (I think there is a missing newline)

File: docs/source/notes/fsdp.rst
Comment (stas00) [reply docs/source/notes/fsdp.rst:74]: oh, it's a good thing we continued that discussion here, since from our discussion on slack I understand that `forward_prefetch=False` is superior memory-wise.

And, so before I update the text, let's quickly confirm that for `backward` it's 1x buffer size?

I guess it'd be useful to add an actual example with numbers and it will also help us to verify that the descriptions are correct. How does this sound:

-------------

Let's consider an example where we have a Transformer-based model sharded over 8 GPUs, the sharding happens at the transformer block-level, and each transformer block contains 0.8B parameters and the parameters are in fp32 (4 bytes each). Which means that once sharded each transformer block will contain 0.1B parameters on each GPU.

* The `forward` pass will communicate in chunks of `1.5*4 = 6GB` in all-gather and it will use a 2x buffer size - so `2*6=12GB` memory usage on GPU.
* In the `backward` pass it will communicate 2 times 6GB each (1x all-gather and 1x reduce-scatter) and will consume a buffer size of 6GB

Thus the total communication buffer size on each GPU will be `12+6=18GB`, regardless of how many transformer blocks there are.

And there will be 3 communications of 6GB per transformer block.

------------------

Did I do the math right? or is the `forward` comms buffer reused for the `backward` pass and thus there is only every 2x buffers - i.e. 12GB total and not 18GB.

Additional questions:
1. what about the embedding - it can be huge - if sharded, would that be an additional buffer
2. if that comms buffer is pretty large, would it more beneficial to wrap at a lower granularity - would that help with a better overlap

Comment (awgu) [reply docs/source/notes/fsdp.rst:74]: Sorry if I already mentioned this, but an important disclaimer is that the current FSDP (and DeepSpeed) use `recordStream` to manage allocations produced in one stream consumed in another, which can lead to more memory usage than expected (compared to our hand derivations). How much more can be "non-deterministic" in that it depends on GPU kernel timing relative to the CPU. The `limit_all_gathers=True` arg is a mitigation to that (and I think [DeepSpeed does this](https://github.com/microsoft/DeepSpeed/blob/05cc3462c948c7bc6484818830133393ff83f824/deepspeed/runtime/zero/partitioned_param_coordinator.py#L301) without option). See @janeyx99 's [note](https://dev-discuss.pytorch.org/t/fsdp-cudacachingallocator-an-outsider-newb-perspective/1486) for some more details.

The new FSDP we are working on addresses this issue but doing so in the existing FSDP is pretty tricky, so we are not planning to land it. As such, we should take the hand derivations with a grain of salt.

---

Regarding your example: Where the `1.5` come from in `1.5*4 = 6GB`? I would have expected it to be `0.8*4 = 3.2GB`.

> let's quickly confirm that for backward it's 1x buffer size?

Yes, in theory it is 1x buffer size, but:
1. The `recordStream` issue is real.
2. The way existing FSDP works with autograd means this is actually like 2x buffer size.
    - Existing FSDP all-gathers the `flat_param`, which is the autograd leaf.
    - It calls `torch.split` to get 1D views into the `flat_param` corresponding to its constituent original parameters.
    - It calls `torch.view` on each 1D split to view back to ND.
    - This means that in backward, we end up with `ViewBackward` (ND -> 1D) and `SplitWithSizesBackward` (which is a concat). In particular, each individual gradient is computed as a separate allocation, and an explicit concat happens to construct the reduce-scatter input buffer. This implies actually a 2x buffer size for reduce-scatter at that peak memory point.

In summary, for backward, it is more like 2x buffer size for reduce-scatter plus any `recordStream` effects.

---

> 1. what about the embedding - it can be huge - if sharded, would that be an additional buffer

Given the rule we discussed that you included in the note starting with "the communication buffer size is determined as follows", we can analyze as follows:
- Suppose we apply FSDP to the root module (e.g. the `Transformer` class). Suppose we further apply FSDP to each transformer block (e.g. the `TransformerBlock` class).
- Most commonly, the embedding and final linear projection are direct children of the root `Transformer` class.
- Following our rule, that means that the embedding and final linear projection are assigned to the root `Transformer`'s flat parameter.
- We have _another_ special rule, which is that the root does not free its parameters after forward because they will be anyways immediately all-gathered in backward.
- Putting this together, this means that the root's flat parameter including the embedding and final projection are all-gathered to begin forward and kept in GPU memory until the end of backward.
- If the embedding and final linear are not weight-tied, then we _could_ further apply FSDP to the embedding and to the final linear. For weight-tied parameters, we require them to be part of the same flat parameter (or else it would get double-counted). That would allow the embedding to be freed after its usage in forward and only all-gathered toward the end of backward.
- Hopefully, this gives a better sense -- each FSDP module gets assigned parameters in its `module.parameters()` except those already assigned to another nested FSDP module, and the FSDP module's `forward()` defines the 'live' interval for its parameters. Hence, the nested `nn.Module` structure can affect the all-gather/free schedule and hence the memory/throughput performance.

> 2. if that comms buffer is pretty large, would it more beneficial to wrap at a lower granularity - would that help with a better overlap

Yes, it could help to wrap at a lower granularity, trading off worse communication efficiency from a smaller message size and from issuing more collectives (incurring the latency linear in world size more times). If activation memory is already pretty dominant, then you may not save too much memory from wrapping at a smaller granularity. It depends on the specific numbers I think.

Comment (stas00) [reply docs/source/notes/fsdp.rst:74]: > Regarding your example: Where the 1.5 come from in 1.5*4 = 6GB? I would have expected it to be 0.8*4 = 3.2GB.

My apologies, I wrote the initial example with larger numbers, then realized that they are too big to be realistic, so I rewrote it to use smaller numbers. So you must have caught it in the middle. Does it checkout now?

Comment (awgu) [reply docs/source/notes/fsdp.rst:74]: > Let's consider an example where we have a Transformer-based model sharded over 8 GPUs, where the sharding happens at the transformer block-level, and each transformer block contains 1.6B parameters and the parameters are in fp32 (4 bytes each). Which means that once sharded each transformer block will contain 0.2B parameters on each GPU.
> 
> * The `forward` pass will communicate in chunks of `0.2*4 = 0.8GB` in all-gather and it will use a 2x buffer size - so `2*0.8=1.6GB` memory usage on GPU.

IIUC, I think you are only counting the all-gather input, which is 1/8 of the all-gather output. However, the all-gather output needs to be allocated on GPU too. For FSDP, these are two separate allocations, so in practice, the buffer size used for one all-gather is `1.6B numel * 4 bytes/numel * (1 + 1/8)` (assuming fp32 all-gather and 8 world size). Since the 1/8 shard is always kept in GPU memory, the additional memory usage is just the `1.6B numel * 4 bytes`. (It may be easier to tabulate by separately counting the 1/8 of the total model size always kept in GPU memory and the dynamically allocated all-gather buffer sizes.) Then, we use 2x of this `1.6B numel * 4 bytes/numel` since we have two alive at once for prefetching.

Comment (stas00) [reply docs/source/notes/fsdp.rst:74]: > the dynamically allocated all-gather buffer sizes

is it really so? won't it be faster to pre-allocate and re-use those buffers - at least for the duration of the transformers blocks - as it won't work with the embeddings.

From what I have seen Deepspeed preallocates all the buffers right away and never changes them. This helps to reduce fragmentation and it's of course faster as it saves alloc/free.

Comment (awgu) [reply docs/source/notes/fsdp.rst:74]: > > the dynamically allocated all-gather buffer sizes
> 
> is it really so? won't it be faster to pre-allocate and re-use those buffers - at least for the duration of the transformers blocks - as it won't work with the embeddings.
> 
> From what I have seen Deepspeed preallocates all the buffers right away and never changes them. This helps to reduce fragmentation and it's of course faster as it saves alloc/free.

I think the allocation time is negligible since it is just some CPU code calling into the caching allocator (assuming that the program has already reached a steady state and no more `cudaMalloc` is called). I agree with the fragmentation issue though.

One design challenge for FSDP is that we use it for internal recommendation model workloads where FSDP does not own the entire model. In such cases, FSDP may only be applied to select modules as part of a larger model with both sparse and dense parts, and it does not necessarily make sense to keep FSDP memory in use throughout the entire forward/backward because it is not always obvious if it will affect the peak memory.

One detail is that since FSDP allocates all-gather and reduce-scatter buffers in separate CUDA streams, and memory blocks in the allocator are tied to a stream, the fragmentation _could_ be somewhat contained to within those stream's memory pools only. That being said, the fragmentation is a real issue, and Zach's expandable segments are one approach to help there.

Comment (awgu) [reply docs/source/notes/fsdp.rst:74]: > yes, you're right, I guess I mixed 2 separate calculations together. We have 2 needs:
> 
> 1. know the comms buffer size, so that the user could run all_reduce benchmark and find the throughput at the comm buffer size and thus possibly choose to adjust what gets wrapped
> 2. know the gpu memory allocations so that they could predict the memory needs
> 
> So the answers are then:
> 
> 1. comms buffer size is `total_transformer_block_params*type_bytes/n_gpus` (regardless of how many blocks there are)
> 2. additional FSDP-protocol GPU memory requirement is in 2 parts:
>    2a. gathered parameters buffer: `total_transformer_block_params*type_bytes` (regardless of how many blocks there are)
>    2b. comms buffer x 2 to hold the currently processed block and the prefetch.
> 
> So in our example:
> 
> 1. `1.6*4/8=0.8GB`
> 
> 2a. `1.6*4=6.4GB` 2b. `0.8*2=1.6GB 2. for a total of 8GB.
> 
> BTW, with the help of the Deepspeed team I created an actual API that can estimate memory usage for different configurations: https://deepspeed.readthedocs.io/en/latest/memory.html#api-to-estimate-memory-usage and the how it gets derived is documented here: https://deepspeed.readthedocs.io/en/latest/memory.html#discussion so perhaps eventually here as well we could quantify the numbers programmatically (possibly only when you rewrite FSDP since as you said earlier above some parts aren't predictable)

Sorry, I am not sure I follow still.
- When we prefetch the all-gather, we need to already allocate the next (prefetched) all-gather's output size. This would be `1.6*4`. I am not sure how to prefetch with only the all-gather's input size, `1.6*4/8`.
- I normally think of the factors that matter for all-gather time as the all-gather output size and the world size. (The all-gather input size can be derived from those two by dividing.) The all-gather output size affects the bandwidth term, and the world size affects the latency term. When you define the comms. buffer size, I am curious why you define it using the all-gather input size (with the `/8` factor)?

Comment (stas00) [reply docs/source/notes/fsdp.rst:74]: > When we prefetch the all-gather, we need to already allocate the next (prefetched) all-gather's output size. This would be 1.6*4. I am not sure how to prefetch with only the all-gather's input size, 1.6*4/8.

I guess we should change the nomenclature - instead of comms buffer - let's call it the comms payload?  so all-gather is called with a payload of `1.6*4/8=0.8GB` per GPU - so now we will look at the output of nccl-tests and match the throughput report for ~0.8GB payload - is that good?

> I normally think of the factors that matter for all-gather time as the all-gather output size and the world size. (The all-gather input size can be derived from those two by dividing.) The all-gather output size affects the bandwidth term, and the world size affects the latency term. When you define the comms. buffer size, I am curious why you define it using the all-gather input size (with the `/8` factor)?

Because if you look at the actual benchmark code - the reported payload in the nccl-tests is per GPU and not per world size. say, in the benchmark each rank all-gathers 4GB of payload (so for 8 gpus it'll be 32GB total), and one will need to look up payload of 4GB and not 32GB in the report.

Does that explain why or am I misunderstanding something?

Here is the pure python implementation of the all-reduce benchmark w/ a hardcoded 4GB payload per rank.
https://github.com/stas00/ml-engineering/blob/master/multi-node/all_reduce_bench.py
its results correspond to nccl-tests's record for 4GB! So that how I knew it's per rank and not per world size.

Comment (awgu) [reply docs/source/notes/fsdp.rst:74]: @stas00 Thanks for the clarification! I was not familiar with the nccl-tests, but what you said makes sense!

Comment (stas00) [reply docs/source/notes/fsdp.rst]: ok, I compiled your awesome design notes into a doc, adding my numerical examples, and splitting the sections into clear sub-topics so hopefully the reader can follow things easily.

I know it has grown a lot since the original doc and you didn't really want to spend time on it, so I apologize again for dragging you into doing some of the work anyway, Andrew.

I propose that as long as I didn't add anything incorrect we can just merge it as is and have future PRs to improve upon this already amazing practical information you shared. The original intention was just to preserve your notes. and for those of us who need to do the math to be able to do it intelligently which these notes enable.

**Comment by stas00** [comment]
The other thing I'm thinking about is just to have a special flag which will just tell the user the comms payload as it happens. Perhaps patching the `all_gather` to also log its payload's `numel` - that way the user will just have an actual number that they can act upon.

File: docs/source/notes/fsdp.rst
Comment (albanD) [inline docs/source/notes/fsdp.rst:89]: You're missing a `:` at the end of this line. That makes everything below a comment and it doesn't render https://docs-preview.pytorch.org/pytorch/pytorch/117323/notes/fsdp.html

File: torch/distributed/fsdp/fully_sharded_data_parallel.py
Comment (albanD) [inline torch/distributed/fsdp/fully_sharded_data_parallel.py:247]: I would not make this a real ".. note:" as there are a lot of them already.
Also I think this is important enough content that we should move it at the top of FSDP doc if @awgu is happy with it.

File: docs/source/notes/fsdp.rst
Comment (stas00) [reply docs/source/notes/fsdp.rst]: thank you, @albanD - fixed

File: torch/distributed/fsdp/fully_sharded_data_parallel.py
Comment (stas00) [reply torch/distributed/fsdp/fully_sharded_data_parallel.py:247]: You suggested on slack to add it to a separate doc with notes. But I'm happy to move it elsewhere if you tell me where. I'm not sure where the FSDP doc is - I didn't find one.

Comment (albanD) [reply torch/distributed/fsdp/fully_sharded_data_parallel.py:247]: Ho no, I meant the link to the note here can be moved at the top of the FSDP docstring so that people will see it :D

Comment (stas00) [reply torch/distributed/fsdp/fully_sharded_data_parallel.py]: Thank you for clarifying, @albanD - I have just pushed that.

**Review by albanD — DISMISSED** [review]
Sounds good from my naive point of view. You can wait for a final approve from @awgu for the content.

**Comment by awgu** [comment]
One quick comment on nccl-tests:
@kwen2501 pointed out to me that for all-gather, it should be using the output size, not the input size that depends on the world size. The code takes the max of the send and expected bytes where the latter is not dependent on world size:
https://github.com/NVIDIA/nccl-tests/blob/1292b25553bd0384f2faa2965f9d82b99797a348/src/common.cu#L602
https://github.com/NVIDIA/nccl-tests/blob/1292b25553bd0384f2faa2965f9d82b99797a348/src/common.cu#L574

File: docs/source/notes/fsdp.rst
Comment (awgu) [inline docs/source/notes/fsdp.rst:115]: One clarification (that we do not need to include) is that, there is an alternate design to this autograd logic that would allow us to free the gathered parameters before concat-ing, allowing us to maintain only using 3x buffer size (2x for all-gather and 1x for gradients or 1x for all-gather and 2x for gradients, changing in a cycle). The rewrite does this.

**Review by awgu — DISMISSED** [review]
LGTM! Thanks again for taking the time!

File: docs/source/notes/fsdp.rst
Comment (awgu) [inline docs/source/notes/fsdp.rst:77]: nit: Is this newline a typo? I think it might be causing the docs build failures:
```
 /var/lib/jenkins/workspace/docs/source/notes/fsdp.rst:78: WARNING: Inline literal start-string without end-string.
```

**Comment by stas00** [comment]
> One quick comment on nccl-tests: @kwen2501 pointed out to me that for all-gather, it should be using the output size, not the input size that depends on the world size. The code takes the max of the send and expected bytes where the latter is not dependent on world size: https://github.com/NVIDIA/nccl-tests/blob/1292b25553bd0384f2faa2965f9d82b99797a348/src/common.cu#L602 https://github.com/NVIDIA/nccl-tests/blob/1292b25553bd0384f2faa2965f9d82b99797a348/src/common.cu#L574

pasting here what you pointed at:
```
  args->sendBytes = sendCount * wordSize(type);
  args->expectedBytes = recvCount * wordSize(type);
  [...]
max(args->sendBytes, args->expectedBytes),
```

both, send and receive appear to depend on the world size.

unless what you were trying to say is that nccl-tests reports the total amount of traffic across all gpus, rather than per gpu.

**Comment by awgu** [comment]
@stas00 I was thinking that the `wordSize` is how many bytes per element, not the world size:
https://github.com/NVIDIA/nccl-tests/blob/1292b25553bd0384f2faa2965f9d82b99797a348/src/common.h#L216

**Comment by stas00** [comment]
ah, it's `word` and not `world`! that makes much more sense.

The explanation for all-gather is here:
https://github.com/NVIDIA/nccl-tests/blob/master/doc/PERFORMANCE.md#allgather

> The AllGather operation requires only to perform the assignment part of the allReduce operation :
> 
>  `o_0 = o_1 = o_2 = ... = o_{n-1} = i_K`
> 
> With K being the rank where the data originates from (K=offset*sendsize).
> 
> The perfect allGather time with a rank bandwidth of B would therefore be :
> 
>  `t = S*(n-1) / (B*n)`
> 
> And the Bus Bandwidth is therefore computed as :
> 
>  `B = S/t * (n-1)/n = algbw * (n-1)/n`
> 
> Note that here, S is the size in bytes of the total array, which for NCCL is equal to `sendcount*sizeof(datatype)*n` as the `sendcount` argument is the count per rank.
> 
> ### Broadcast
> 
> The broadcast operation representation is similar to allGather :
> 
>  `o_0 = o_1 = o_2 = ... = o_{n-1} = i_R`
> 
> R being the root of the operation.
> 
> However, in this case, since the i_R input is not evenly distributed on the ranks, we cannot use all N links to perform the transfer operations.
> Indeed, *all* data has to get out of the root rank, hence the bottleneck is on the root rank which only has B as capacity to get data out :
> 
>  `t = S/B`
> 
> And :
> 
>  `B = S/t`
> 

OK, is @kwen2501 suggesting that nccl-tests logic for all-gather is incorrect?

**Comment by stas00** [comment]
and the rest of collectives is correct then? I usually just use the all_reduce benchmark when testing whether the setup is correct.

File: docs/source/notes/fsdp.rst
Comment (stas00) [reply docs/source/notes/fsdp.rst:115]: That's wonderful news, Andrew!

File: docs/source/notes/fsdp.rst
Comment (stas00) [reply docs/source/notes/fsdp.rst:77]: oh, it's complaining about the next para starting with a lower case letter I think. Let's see if the mended version solves this.

**Comment by awgu** [comment]
> OK, is @kwen2501 suggesting that nccl-tests logic for all-gather is incorrect?

@stas00 I can let @kwen2501 comment since I have not personally worked with or looked into the nccl-tests much.

IIUC, the question is what number should users be using when using nccl-tests to estimate the bus bandwidth. What @kwen2501 pointed out should be that the line printed will use `max(args->sendBytes, args->expectedBytes)`, which in the case of all-gather will be the all-gather output size (without the `1/N` factor for `N` GPUs).

I do not think that this conflicts with what is outlined in the nccl-tests' [readme for all-gather](https://github.com/NVIDIA/nccl-tests/blob/master/doc/PERFORMANCE.md#allgather). 

> and the rest of collectives is correct then? I usually just use the all_reduce benchmark when testing whether the setup is correct.

Again, speaking without experience with nccl-tests, I think that using all-reduce would be an approximation for all-gather since all-reduce can have logarithmic latency (under the tree algorithm), whereas all-gather today has linear latency (since it only supports a ring algorithm). For large scales (e.g. >=128 GPUs), we do find this latency to be non-negligible, in which case, it might not be as suitable to use all-reduce to approximate all-gather.

Let me know if I am misunderstanding anything in either part!

**Comment by stas00** [comment]
What I meant to say that I use one consistent `all_reduce` benchmark to make sure that I get within ~80% of the advertised peak throughput on a 4GB payload - that way I know the vendor gave me the right hardware, and I have configured the software correctly.

Then I don't really have any choice to how fast any of the collectives runs because that's beyond my control, other than perhaps mucking around with NCCL env vars to choose some specific protocols.

So really as an advanced user I just need to know which payload row in the nccl-tests output I need to look at and then I can calculate the overhead of the comms. Ideally I'd look at them as about the same, so only checking one of them. This is all an approximation anyway, to tell whether the network is fast enough not to be a huge bottleneck.

As you know ZeRO (Deepspeed or FSDP) needs more bandwidth than 3D-parallelism, albeit with the hybrid solution if the model can fit into a single node, it's less of a difference. But there are still providers out there who give like 50Gbps inter-node connectivity - so ZeRO users will experience lower TFLOPs.

When we trained IDEFICS-80B we got to barely 90TFLOPS using Deepspeed ZeRO-3 on 340Gbps effective internode, while doing BLOOM-176B we got 150TFLOPS with just 150Gbps - same A100s otherwise. So the network impact is huge for ZeRO-based protocols, because of the 2 additional gathers.

**Comment by stas00** [comment]
I tried my best to document my understanding here:
https://github.com/stas00/ml-engineering/tree/master/network#understanding-why-inter-node-network-speed-is-of-a-huge-importance
If you're inspired to read and let me know if I made mistakes I'm all ears.

**Comment by awgu** [comment]
@stas00 Got it! Thanks for the detailed explanation :)

**Comment by awgu** [comment]
I missed the docs build breakage when I did `@pytorchbot merge -i` -- reverting now.

**Comment by stas00** [comment]
Hmm, I wonder why linter doesn't catch those - had a single "`" instead of "``" in 2 places. Hopefully it'll be happy now.

**Review by awgu — APPROVED** [review]
Thanks! 

I will double check the docs test before landing this time 😅 .

**Comment by awgu** [comment]
These failing tests look unrelated:
- [pull / linux-focal-py3.11-clang10 / test (dynamo, 1, 3, linux.2xlarge)](https://hud.pytorch.org/pr/pytorch/pytorch/117323#20707504065) ([gh](https://github.com/pytorch/pytorch/actions/runs/7604350691/job/20707504065))
test_weak.py::WeakTest::test_make_weak_keyed_dict_from_weak_keyed_dict
- [pull / linux-focal-py3.8-clang10 / test (dynamo, 1, 3, linux.2xlarge)](https://hud.pytorch.org/pr/pytorch/pytorch/117323#20707503535) ([gh](https://github.com/pytorch/pytorch/actions/runs/7604350691/job/20707503535))
test_weak.py::WeakTest::test_make_weak_keyed_dict_from_weak_keyed_dict

The others are already determined as flaky:
- [inductor / rocm5.7-py3.8-inductor / test (inductor, 1, 1, linux.rocm.gpu.2)](https://hud.pytorch.org/pr/pytorch/pytorch/117323#20707486388) ([gh](https://github.com/pytorch/pytorch/actions/runs/7604350960/job/20707486388))
- [pull / linux-focal-py3-clang9-android-ndk-r21e-gradle-custom-build-single-full-jit / build-and-test (default, 1, 1, linux.2xlarge)](https://hud.pytorch.org/pr/pytorch/pytorch/117323#20707338505) ([gh](https://github.com/pytorch/pytorch/actions/runs/7604350691/job/20707338505))
- [trunk / linux-focal-rocm5.7-py3.8 / test (default, 1, 1, linux.rocm.gpu)](https://hud.pytorch.org/pr/pytorch/pytorch/117323#20707485232) ([gh](https://github.com/pytorch/pytorch/actions/runs/7604350868/job/20707485232))

**Comment by awgu** [comment]
These docs tests are all passing:
- pull / linux-docs / build-docs-cpp-false (pull_request)
- pull / linux-docs / build-docs-python-false (pull_request)
- pull / linux-docs / build-docs-functorch-false (pull_request)
- pull / linux-jammy-py3.8-gcc11 / test (docs_test, 1, 1, linux.2xlarge) (pull_request)

These were the only ones I saw with "docs" in their name.

**Pull request closed without merging**
